File: wire-schema/src/test/java/com/squareup/wire/internal/protoparser/ProtoFileElementTest.java
/*
 * Copyright (C) 2013 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.internal.protoparser;

import com.squareup.wire.internal.protoparser.OptionElement.Kind;
import com.squareup.wire.schema.Location;
import java.util.Arrays;
import java.util.Collections;
import org.junit.Test;

import static com.squareup.wire.internal.protoparser.ProtoFileElement.MAX_TAG_VALUE;
import static com.squareup.wire.internal.protoparser.ProtoFileElement.MIN_TAG_VALUE;
import static com.squareup.wire.internal.protoparser.ProtoFileElement.Syntax.PROTO_2;
import static com.squareup.wire.internal.protoparser.ProtoFileElement.isValidTag;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.fail;

public class ProtoFileElementTest {
  Location location = Location.get("file.proto");

  @Test public void nullBuilderValuesThrow() {
    try {
      ProtoFileElement.builder((Location) null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("location");
    }
    try {
      ProtoFileElement.builder(location).packageName(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("packageName");
    }
    try {
      ProtoFileElement.builder(location).syntax(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("syntax");
    }
    try {
      ProtoFileElement.builder(location).addDependency(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("dependency");
    }
    try {
      ProtoFileElement.builder(location).addDependencies(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("dependencies");
    }
    try {
      ProtoFileElement.builder(location).addDependencies(Collections.<String>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("dependency");
    }
    try {
      ProtoFileElement.builder(location).addPublicDependency(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("dependency");
    }
    try {
      ProtoFileElement.builder(location).addPublicDependencies(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("dependencies");
    }
    try {
      ProtoFileElement.builder(location)
          .addPublicDependencies(Collections.<String>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("dependency");
    }
    try {
      ProtoFileElement.builder(location).addType(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("type");
    }
    try {
      ProtoFileElement.builder(location).addTypes(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("types");
    }
    try {
      ProtoFileElement.builder(location).addTypes(Collections.<TypeElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("type");
    }
    try {
      ProtoFileElement.builder(location).addService(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("service");
    }
    try {
      ProtoFileElement.builder(location).addServices(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("services");
    }
    try {
      ProtoFileElement.builder(location)
          .addServices(Collections.<ServiceElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("service");
    }
    try {
      ProtoFileElement.builder(location).addExtendDeclaration(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("extend");
    }
    try {
      ProtoFileElement.builder(location).addExtendDeclarations(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("extendDeclarations");
    }
    try {
      ProtoFileElement.builder(location).addExtendDeclarations(
          Collections.<ExtendElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("extend");
    }
    try {
      ProtoFileElement.builder(location).addOption(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("option");
    }
    try {
      ProtoFileElement.builder(location).addOptions(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("options");
    }
    try {
      ProtoFileElement.builder(location)
          .addOptions(Collections.<OptionElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("option");
    }
  }

  @Test public void tagValueValidation() {
    assertThat(isValidTag(MIN_TAG_VALUE - 1)).isFalse(); // Less than minimum.
    assertThat(isValidTag(MIN_TAG_VALUE)).isTrue();
    assertThat(isValidTag(1234)).isTrue();
    assertThat(isValidTag(19222)).isFalse(); // Reserved range.
    assertThat(isValidTag(2319573)).isTrue();
    assertThat(isValidTag(MAX_TAG_VALUE)).isTrue();
    assertThat(isValidTag(MAX_TAG_VALUE + 1)).isFalse(); // Greater than maximum.
  }

  @Test public void emptyToSchema() {
    ProtoFileElement file = ProtoFileElement.builder(location).build();
    String expected = "// file.proto\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }

  @Test public void emptyWithPackageToSchema() {
    ProtoFileElement file = ProtoFileElement.builder(location)
        .packageName("example.simple")
        .build();
    String expected = ""
        + "// file.proto\n"
        + "package example.simple;\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }

  @Test public void simpleToSchema() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    ProtoFileElement file = ProtoFileElement.builder(location).addType(element).build();
    String expected = ""
        + "// file.proto\n"
        + "\n"
        + "message Message {}\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }

  @Test public void simpleWithImportsToSchema() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    ProtoFileElement file = ProtoFileElement.builder(location)
        .addDependency("example.other")
        .addType(element)
        .build();
    String expected = ""
        + "// file.proto\n"
        + "\n"
        + "import \"example.other\";\n"
        + "\n"
        + "message Message {}\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleDependencies() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    ProtoFileElement file = ProtoFileElement.builder(location)
        .addDependencies(Arrays.asList("example.other", "example.another"))
        .addType(element)
        .build();
    assertThat(file.dependencies()).hasSize(2);
  }

  @Test public void simpleWithPublicImportsToSchema() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    ProtoFileElement file = ProtoFileElement.builder(location)
        .addPublicDependency("example.other")
        .addType(element)
        .build();
    String expected = ""
        + "// file.proto\n"
        + "\n"
        + "import public \"example.other\";\n"
        + "\n"
        + "message Message {}\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultiplePublicDependencies() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    ProtoFileElement file = ProtoFileElement.builder(location)
        .addPublicDependencies(Arrays.asList("example.other", "example.another"))
        .addType(element)
        .build();
    assertThat(file.publicDependencies()).hasSize(2);
  }

  @Test public void simpleWithBothImportsToSchema() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    ProtoFileElement file = ProtoFileElement.builder(location)
        .addDependency("example.thing")
        .addPublicDependency("example.other")
        .addType(element)
        .build();
    String expected = ""
        + "// file.proto\n"
        + "\n"
        + "import \"example.thing\";\n"
        + "import public \"example.other\";\n"
        + "\n"
        + "message Message {}\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }

  @Test public void simpleWithServicesToSchema() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    ServiceElement service = ServiceElement.builder(location).name("Service").build();
    ProtoFileElement
        file = ProtoFileElement.builder(location).addType(element).addService(service).build();
    String expected = ""
        + "// file.proto\n"
        + "\n"
        + "message Message {}\n"
        + "\n"
        + "service Service {}\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleServices() {
    ServiceElement service1 = ServiceElement.builder(location).name("Service1").build();
    ServiceElement service2 = ServiceElement.builder(location).name("Service2").build();
    ProtoFileElement file = ProtoFileElement.builder(location)
        .addServices(Arrays.asList(service1, service2))
        .build();
    assertThat(file.services()).hasSize(2);
  }

  @Test public void simpleWithOptionsToSchema() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    OptionElement option = OptionElement.create("kit", Kind.STRING, "kat");
    ProtoFileElement
        file = ProtoFileElement.builder(location).addOption(option).addType(element).build();
    String expected = ""
        + "// file.proto\n"
        + "\n"
        + "option kit = \"kat\";\n"
        + "\n"
        + "message Message {}\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleOptions() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    OptionElement kitKat = OptionElement.create("kit", Kind.STRING, "kat");
    OptionElement fooBar = OptionElement.create("foo", Kind.STRING, "bar");
    ProtoFileElement file = ProtoFileElement.builder(location)
        .addOptions(Arrays.asList(kitKat, fooBar))
        .addType(element)
        .build();
    assertThat(file.options()).hasSize(2);
  }

  @Test public void simpleWithExtendsToSchema() {
    ProtoFileElement file = ProtoFileElement.builder(location)
        .addExtendDeclaration(ExtendElement.builder(location.at(5, 1)).name("Extend").build())
        .addType(MessageElement.builder(location).name("Message").build())
        .build();
    String expected = ""
        + "// file.proto\n"
        + "\n"
        + "message Message {}\n"
        + "\n"
        + "extend Extend {}\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleExtends() {
    ExtendElement extend1 = ExtendElement.builder(location).name("Extend1").build();
    ExtendElement extend2 = ExtendElement.builder(location).name("Extend2").build();
    ProtoFileElement file = ProtoFileElement.builder(location)
        .addExtendDeclarations(Arrays.asList(extend1, extend2))
        .build();
    assertThat(file.extendDeclarations()).hasSize(2);
  }

  @Test public void multipleEverythingToSchema() {
    TypeElement element1 = MessageElement.builder(location.at(10, 1))
        .name("Message1")
        .qualifiedName("example.simple.Message1")
        .build();
    TypeElement element2 = MessageElement.builder(location.at(11, 1))
        .name("Message2")
        .qualifiedName("example.simple.Message2")
        .build();
    ExtendElement extend1 = ExtendElement.builder(location.at(13, 1))
        .name("Extend1")
        .qualifiedName("example.simple.Extend1")
        .build();
    ExtendElement extend2 = ExtendElement.builder(location.at(14, 1))
        .name("Extend2")
        .qualifiedName("example.simple.Extend2")
        .build();
    OptionElement option1 = OptionElement.create("kit", Kind.STRING, "kat");
    OptionElement option2 = OptionElement.create("foo", Kind.STRING, "bar");
    ServiceElement service1 = ServiceElement.builder(location.at(16, 1))
        .name("Service1")
        .qualifiedName("example.simple.Service1")
        .build();
    ServiceElement service2 = ServiceElement.builder(location.at(17, 1))
        .name("Service2")
        .qualifiedName("example.simple.Service2")
        .build();
    ProtoFileElement file = ProtoFileElement.builder(location)
        .packageName("example.simple")
        .addDependency("example.thing")
        .addPublicDependency("example.other")
        .addType(element1)
        .addType(element2)
        .addService(service1)
        .addService(service2)
        .addExtendDeclaration(extend1)
        .addExtendDeclaration(extend2)
        .addOption(option1)
        .addOption(option2)
        .build();
    String expected = ""
        + "// file.proto\n"
        + "package example.simple;\n"
        + "\n"
        + "import \"example.thing\";\n"
        + "import public \"example.other\";\n"
        + "\n"
        + "option kit = \"kat\";\n"
        + "option foo = \"bar\";\n"
        + "\n"
        + "message Message1 {}\n"
        + "message Message2 {}\n"
        + "\n"
        + "extend Extend1 {}\n"
        + "extend Extend2 {}\n"
        + "\n"
        + "service Service1 {}\n"
        + "service Service2 {}\n";
    assertThat(file.toSchema()).isEqualTo(expected);

    // Re-parse the expected string into a ProtoFile and ensure they're equal.
    ProtoFileElement parsed = ProtoParser.parse(location, expected);
    assertThat(parsed).isEqualTo(file);
  }

  @Test public void syntaxToSchema() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    ProtoFileElement
        file = ProtoFileElement.builder(location).syntax(PROTO_2).addType(element).build();
    String expected = ""
        + "// file.proto\n"
        + "syntax \"proto2\";\n"
        + "\n"
        + "message Message {}\n";
    assertThat(file.toSchema()).isEqualTo(expected);
  }
}

File: wire-library/wire-schema/src/jvmTest/kotlin/com/squareup/wire/schema/internal/parser/ProtoFileElementTest.kt
/*
 * Copyright (C) 2013 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema.internal.parser

import com.squareup.wire.schema.Field
import com.squareup.wire.schema.Location
import com.squareup.wire.schema.ProtoFile.Syntax.PROTO_2
import com.squareup.wire.schema.ProtoFile.Syntax.PROTO_3
import com.squareup.wire.schema.internal.parser.OptionElement.Companion.PACKED_OPTION_ELEMENT
import com.squareup.wire.schema.internal.parser.OptionElement.Kind
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class ProtoFileElementTest {
  internal var location = Location.get("some/folder", "file.proto")

  @Test
  fun emptyToSchema() {
    val file = ProtoFileElement(location = location)
    val expected = "// file.proto\n"
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun emptyWithPackageToSchema() {
    val file = ProtoFileElement(
        location = location,
        packageName = "example.simple"
    )
    val expected = """
        |// file.proto
        |package example.simple;
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun simpleToSchema() {
    val element = MessageElement(
        location = location,
        name = "Message"
    )
    val file = ProtoFileElement(
        location = location,
        types = listOf(element)
    )
    val expected = """
        |// file.proto
        |
        |message Message {}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun simpleWithImportsToSchema() {
    val element = MessageElement(
        location = location,
        name = "Message"
    )
    val file = ProtoFileElement(
        location = location,
        imports = listOf("example.other"),
        types = listOf(element)
    )
    val expected = """
        |// file.proto
        |
        |import "example.other";
        |
        |message Message {}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleDependencies() {
    val element = MessageElement(
        location = location,
        name = "Message"
    )
    val file = ProtoFileElement(
        location = location,
        imports = listOf("example.other", "example.another"),
        types = listOf(element)
    )
    assertThat(file.imports).hasSize(2)
  }

  @Test
  fun simpleWithPublicImportsToSchema() {
    val element = MessageElement(
        location = location,
        name = "Message"
    )
    val file = ProtoFileElement(
        location = location,
        publicImports = listOf("example.other"),
        types = listOf(element)
    )
    val expected = """
        |// file.proto
        |
        |import public "example.other";
        |
        |message Message {}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultiplePublicDependencies() {
    val element = MessageElement(
        location = location,
        name = "Message"
    )
    val file = ProtoFileElement(location = location,
        publicImports = listOf("example.other", "example.another"),
        types = listOf(element)
    )
    assertThat(file.publicImports).hasSize(2)
  }

  @Test
  fun simpleWithBothImportsToSchema() {
    val element = MessageElement(
        location = location,
        name = "Message"
    )
    val file = ProtoFileElement(location = location,
        imports = listOf("example.thing"),
        publicImports = listOf("example.other"),
        types = listOf(element)
    )
    val expected = """
        |// file.proto
        |
        |import "example.thing";
        |import public "example.other";
        |
        |message Message {}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun simpleWithServicesToSchema() {
    val element = MessageElement(
        location = location,
        name = "Message"
    )
    val service = ServiceElement(
        location = location,
        name = "Service"
    )
    val file = ProtoFileElement(
        location = location,
        types = listOf(element),
        services = listOf(service)
    )
    val expected = """
        |// file.proto
        |
        |message Message {}
        |
        |service Service {}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleServices() {
    val service1 = ServiceElement(
        location = location,
        name = "Service1"
    )
    val service2 = ServiceElement(
        location = location,
        name = "Service2"
    )
    val file = ProtoFileElement(
        location = location,
        services = listOf(service1, service2)
    )
    assertThat(file.services).hasSize(2)
  }

  @Test
  fun simpleWithOptionsToSchema() {
    val element = MessageElement(
        location = location,
        name = "Message"
    )
    val option = OptionElement.create("kit", Kind.STRING, "kat")
    val file = ProtoFileElement(
        location = location,
        options = listOf(option),
        types = listOf(element)
    )
    val expected = """
        |// file.proto
        |
        |option kit = "kat";
        |
        |message Message {}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleOptions() {
    val element = MessageElement(
        location = location,
        name = "Message"
    )
    val kitKat = OptionElement.create("kit", Kind.STRING, "kat")
    val fooBar = OptionElement.create("foo", Kind.STRING, "bar")
    val file = ProtoFileElement(
        location = location,
        options = listOf(kitKat, fooBar),
        types = listOf(element)
    )
    assertThat(file.options).hasSize(2)
  }

  @Test
  fun simpleWithExtendsToSchema() {
    val file = ProtoFileElement(
        location = location,
        extendDeclarations = listOf(ExtendElement(location = location.at(5, 1), name = "Extend")),
        types = listOf(MessageElement(location = location, name = "Message"))
    )
    val expected = """
        |// file.proto
        |
        |message Message {}
        |
        |extend Extend {}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleExtends() {
    val extend1 = ExtendElement(location = location, name = "Extend1")
    val extend2 = ExtendElement(location = location, name = "Extend2")
    val file = ProtoFileElement(
        location = location,
        extendDeclarations = listOf(extend1, extend2)
    )
    assertThat(file.extendDeclarations).hasSize(2)
  }

  @Test
  fun multipleEverythingToSchema() {
    val element1 = MessageElement(location = location.at(10, 1), name = "Message1")
    val element2 = MessageElement(location = location.at(11, 1), name = "Message2")
    val extend1 = ExtendElement(location = location.at(13, 1), name = "Extend1")
    val extend2 = ExtendElement(location = location.at(14, 1), name = "Extend2")
    val option1 = OptionElement.create("kit", Kind.STRING, "kat")
    val option2 = OptionElement.create("foo", Kind.STRING, "bar")
    val service1 = ServiceElement(
        location = location.at(16, 1),
        name = "Service1"
    )
    val service2 = ServiceElement(
        location = location.at(17, 1),
        name = "Service2"
    )
    val file = ProtoFileElement(
        location = location,
        packageName = "example.simple",
        imports = listOf("example.thing"),
        publicImports = listOf("example.other"),
        types = listOf(element1, element2),
        services = listOf(service1, service2),
        extendDeclarations = listOf(extend1, extend2),
        options = listOf(option1, option2)
    )
    val expected = """
        |// file.proto
        |package example.simple;
        |
        |import "example.thing";
        |import public "example.other";
        |
        |option kit = "kat";
        |option foo = "bar";
        |
        |message Message1 {}
        |message Message2 {}
        |
        |extend Extend1 {}
        |extend Extend2 {}
        |
        |service Service1 {}
        |service Service2 {}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)

    // Re-parse the expected string into a ProtoFile and ensure they're equal.
    val parsed = ProtoParser.parse(location, expected)
    assertThat(parsed).isEqualTo(file)
  }

  @Test
  fun syntaxToSchema() {
    val element = MessageElement(location = location, name = "Message")
    val file = ProtoFileElement(
        location = location,
        syntax = PROTO_2,
        types = listOf(element)
    )
    val expected = """
        |// file.proto
        |syntax = "proto2";
        |
        |message Message {}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)
  }

  @Test
  fun defaultIsSetInProto2() {
    val field = FieldElement(
        location = location.at(9, 3),
        label = Field.Label.REQUIRED,
        type = "string",
        name = "name",
        tag = 1,
        defaultValue = "defaultValue"
    )
    val message =
        MessageElement(location = location.at(8, 1), name = "Message", fields = listOf(field))
    val file = ProtoFileElement(
        syntax = PROTO_2,
        location = location,
        packageName = "example.simple",
        imports = listOf("example.thing"),
        publicImports = listOf("example.other"),
        types = listOf(message)
    )
    val expected = """
        |// file.proto
        |syntax = "proto2";
        |package example.simple;
        |
        |import "example.thing";
        |import public "example.other";
        |
        |message Message {
        |  required string name = 1 [default = "defaultValue"];
        |}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)

    // Re-parse the expected string into a ProtoFile and ensure they're equal.
    val parsed = ProtoParser.parse(location, expected)
    assertThat(parsed).isEqualTo(file)
  }

  @Test
  fun defaultIsNotSetInProto3() {
    val field = FieldElement(
        location = location.at(9, 3),
        label = Field.Label.REQUIRED,
        type = "string",
        name = "name",
        tag = 1,
        defaultValue = "defaultValue"
    )
    val message =
        MessageElement(location = location.at(8, 1), name = "Message", fields = listOf(field))
    val file = ProtoFileElement(
        syntax = PROTO_3,
        location = location,
        packageName = "example.simple",
        imports = listOf("example.thing"),
        publicImports = listOf("example.other"),
        types = listOf(message)
    )
    assertThat(file.toSchema()).doesNotContain("defaultValue")

    // TODO(benoit|masaru): Ignore the test below now. This will be fixed by #1386.
    // val expected = """
    //     |// file.proto
    //     |syntax = "proto3";
    //     |package example.simple;
    //     |
    //     |import "example.thing";
    //     |import public "example.other";
    //     |
    //     |message Message {
    //     |  string name = 1;
    //     |}
    //     |""".trimMargin()
    // assertThat(file.toSchema()).isEqualTo(expected)
    // // Re-parse the expected string into a ProtoFile and ensure they're equal.
    // val parsed = ProtoParser.parse(location, expected)
    // assertThat(parsed).isEqualTo(file)
  }

  @Test
  fun convertPackedOptionFromWireSchemaInProto2() {
    val fieldNumeric = FieldElement(
        location = location.at(6, 3),
        label = Field.Label.REPEATED,
        type = "int32",
        name = "numeric_without_packed_option",
        tag = 1
    )
    val fieldNumericPackedTrue = FieldElement(
        location = location.at(7, 3),
        label = Field.Label.REPEATED,
        type = "int32",
        name = "numeric_packed_true",
        tag = 2,
        options = listOf(PACKED_OPTION_ELEMENT)
    )
    val fieldNumericPackedFalse = FieldElement(
        location = location.at(8, 3),
        label = Field.Label.REPEATED,
        type = "int32",
        name = "numeric_packed_false",
        tag = 3,
        options = listOf(PACKED_OPTION_ELEMENT.copy(value = "false"))
    )
    val fieldString = FieldElement(
        location = location.at(9, 3),
        label = Field.Label.REPEATED,
        type = "string",
        name = "string_without_packed_option",
        tag = 4
    )
    val fieldStringPackedTrue = FieldElement(
        location = location.at(10, 3),
        label = Field.Label.REPEATED,
        type = "string",
        name = "string_packed_true",
        tag = 5,
        options = listOf(PACKED_OPTION_ELEMENT)
    )
    val fieldStringPackedFalse = FieldElement(
        location = location.at(11, 3),
        label = Field.Label.REPEATED,
        type = "string",
        name = "string_packed_false",
        tag = 6,
        options = listOf(PACKED_OPTION_ELEMENT.copy(value = "false"))
    )

    val message = MessageElement(
        location = location.at(5, 1),
        name = "Message",
        fields = listOf(fieldNumeric, fieldNumericPackedTrue, fieldNumericPackedFalse, fieldString,
            fieldStringPackedTrue, fieldStringPackedFalse))
    val file = ProtoFileElement(
        syntax = PROTO_2,
        location = location,
        packageName = "example.simple",
        imports = emptyList(),
        publicImports = emptyList(),
        types = listOf(message)
    )
    val expected = """
        |// file.proto
        |syntax = "proto2";
        |package example.simple;
        |
        |message Message {
        |  repeated int32 numeric_without_packed_option = 1;
        |  repeated int32 numeric_packed_true = 2 [packed = true];
        |  repeated int32 numeric_packed_false = 3 [packed = false];
        |  repeated string string_without_packed_option = 4;
        |  repeated string string_packed_true = 5 [packed = true];
        |  repeated string string_packed_false = 6 [packed = false];
        |}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)

    // Re-parse the expected string into a ProtoFile and ensure they're equal.
    val parsed = ProtoParser.parse(location, expected)
    assertThat(parsed).isEqualTo(file)
  }

  @Test
  fun convertPackedOptionFromWireSchemaInProto3() {
    val fieldNumeric = FieldElement(
        location = location.at(6, 3),
        label = Field.Label.REPEATED,
        type = "int32",
        name = "numeric_without_packed_option",
        tag = 1
    )
    val fieldNumericPackedTrue = FieldElement(
        location = location.at(7, 3),
        label = Field.Label.REPEATED,
        type = "int32",
        name = "numeric_packed_true",
        tag = 2,
        options = listOf(PACKED_OPTION_ELEMENT)
    )
    val fieldNumericPackedFalse = FieldElement(
        location = location.at(8, 3),
        label = Field.Label.REPEATED,
        type = "int32",
        name = "numeric_packed_false",
        tag = 3,
        options = listOf(PACKED_OPTION_ELEMENT.copy(value = "false"))
    )
    val fieldString = FieldElement(
        location = location.at(9, 3),
        label = Field.Label.REPEATED,
        type = "string",
        name = "string_without_packed_option",
        tag = 4
    )
    val fieldStringPackedTrue = FieldElement(
        location = location.at(10, 3),
        label = Field.Label.REPEATED,
        type = "string",
        name = "string_packed_true",
        tag = 5,
        options = listOf(PACKED_OPTION_ELEMENT)
    )
    val fieldStringPackedFalse = FieldElement(
        location = location.at(11, 3),
        label = Field.Label.REPEATED,
        type = "string",
        name = "string_packed_false",
        tag = 6,
        options = listOf(PACKED_OPTION_ELEMENT.copy(value = "false"))
    )

    val message = MessageElement(
        location = location.at(5, 1),
        name = "Message",
        fields = listOf(fieldNumeric, fieldNumericPackedTrue, fieldNumericPackedFalse, fieldString,
            fieldStringPackedTrue, fieldStringPackedFalse)
    )
    val file = ProtoFileElement(
        syntax = PROTO_3,
        location = location,
        packageName = "example.simple",
        imports = emptyList(),
        publicImports = emptyList(),
        types = listOf(message)
    )
    val expected = """
        |// file.proto
        |syntax = "proto3";
        |package example.simple;
        |
        |message Message {
        |  repeated int32 numeric_without_packed_option = 1;
        |  repeated int32 numeric_packed_true = 2 [packed = true];
        |  repeated int32 numeric_packed_false = 3 [packed = false];
        |  repeated string string_without_packed_option = 4;
        |  repeated string string_packed_true = 5 [packed = true];
        |  repeated string string_packed_false = 6 [packed = false];
        |}
        |""".trimMargin()
    assertThat(file.toSchema()).isEqualTo(expected)

    // Re-parse the expected string into a ProtoFile and ensure they're equal.
    val parsed = ProtoParser.parse(location, expected)
    assertThat(parsed).isEqualTo(file)
  }
}
